index.js

```diff
@@ -0,0 +1,979 @@
+          var result = {};
+          for (var k in this) result[k] = this[k];
+          for (k in strMap) result[k] = strMap[k];
+          return result;
+        },
+
+        //  Object#map :: StrMap a ~> (a -> b) -> StrMap b
+        'fantasy-land/map': function(f) {
+          var result = {};
+          for (var k in this) result[k] = f(this[k]);
+          return result;
+        },
+
+        //  Object#reduce :: StrMap a ~> ((b, a) -> b, b) -> b
+        'fantasy-land/reduce': function(f, initial) {
+          var result = initial;
+          for (var k in this) result = f(result, this[k]);
+          return result;
+        }
+      }
+    },
+
+    Function: {
+      prototype: {
+        //  Function#toString :: Function ~> () -> String
+        toString: function() { return String(this); },
+
+        //  Function#equals :: Function ~> Function -> Boolean
+        'fantasy-land/equals': function(other) { return other === this; },
+
+        //  Function#map :: (a -> b) ~> (b -> c) -> (a -> c)
+        'fantasy-land/map': function(f) {
+          var functor = this;
+          return function(x) { return f(functor(x)); };
+        },
+
+        //  Function#promap :: (b -> c) ~> (a -> b, c -> d) -> (a -> d)
+        'fantasy-land/promap': function(before, after) {
+          var profunctor = this;
+          return function(x) { return after(profunctor(before(x))); };
+        },
+
+        //  Function#ap :: (a -> b) ~> (a -> b -> c) -> (a -> c)
+        'fantasy-land/ap': function(a2b2c) {
+          var a2b = this;
+          return function(x) { return a2b2c(x)(a2b(x)); };
+        }
+      }
+    }
+
+  };
+
+  //  $seenStack :: Array (Array Any)
+  //
+  //  The strategy for handling circular references is as follows:
+  //
+  //  1.  ‘toString’ pushes a new empty array onto $seenStack.
+  //
+  //  2.  ‘toString’ calls the private ‘show’ function inside a ‘try’ block.
+  //
+  //  3.  ‘show’ binds the last (most recently added) element of $seenStack
+  //      to a local variable named $seen.
+  //
+  //  4.  ‘show’ determines whether the value to be shown appears in $seen,
+  //      using reference equality. If it is, a circular reference has been
+  //      detected, and ‘show’ returns '<Circular>'. Otherwise, it continues.
+  //
+  //  5.  ‘show’ pushes the value to be shown onto $seen.
+  //
+  //  6.  ‘show’ dispatches to a ‘toString’ method inside a ‘try’ block.
+  //
+  //  7.  The ‘finally’ block in ‘show’ pops the value pushed onto $seen in
+  //      step 5.
+  //
+  //  8.  If the method invocation in step 6 succeeded, its result is returned.
+  //      Otherwise, the exception is allowed to propagate.
+  //
+  //  9.  The ‘finally’ block in ‘toString’ pops the array pushed onto
+  //      $seenStack in step 1.
+  //
+  //  10. If the call to ‘show’ in step 2 succeeded, its result is returned.
+  //      Otherwise, the exception is allowed to propagate.
+  var $seenStack = [];
```

Comment:
Thank you for performing such a thorough investigation! I had a feeling you were the right person for the job. :)

> I will now ponder whether this is a problem, and what could be a solution.

I believe it's a problem, but I don't see a way to fix it without an API change. An elegant solution is to provide an `Any -> String` function as an argument to the method. At that point the method would be sufficiently different from a conventional `toString` method that it should probably have a different name; perhaps `show`. Let's consider `Identity#show`:

```javascript
//  Identity#show :: Identity a ~> (b -> String) -> String
Identity.prototype.show = function(show) {
  return 'Identity(' + show(this.value) + ')';
};
```

_Some time later:_ I think there *is* a way to handle the case you discovered without changing the API. Instead of looking for the value in the *last* array in `$seenStack`, we look to see whether it is in *any* array in `$seenStack`. This fixes your case without breaking any of the existing test cases, but I haven't yet convinced myself that it's sound. Could you pull the latest commit and try to break it again? :)